File: packages/tasklens/src/utils/time.ts
// src/utils/time.ts

let now = Date.now;

/**
 * Returns the current timestamp in milliseconds.
 * Can be mocked for testing purposes.
 */
export function getCurrentTimestamp(): number {
  return now();
}

/**
 * Mocks the current timestamp for testing.
 * @param timestamp The timestamp to return, or a function that returns it.
 */
export function mockCurrentTimestamp(timestamp: number | (() => number)): void {
  if (typeof timestamp === 'number') {
    now = () => timestamp;
  } else {
    now = timestamp;
  }
}

/**
 * Resets the current timestamp mock.
 */
export function resetCurrentTimestampMock(): void {
  now = Date.now;
}

/**
 * Converts days to milliseconds.
 * @param days Number of days.
 */
export function daysToMilliseconds(days: number): number {
  return days * 24 * 60 * 60 * 1000;
}

/**
 * Converts hours to milliseconds.
 * @param hours Number of hours.
 */
export function hoursToMilliseconds(hours: number): number {
  return hours * 60 * 60 * 1000;
}

/**
 * Helper: Convert frequency/interval to milliseconds.
 * Note: simplified logic for MVP (ignoring leap years/DST nuances for basic "days/weeks").
 */
export function getIntervalMs(
  frequency: 'minutes' | 'hours' | 'daily' | 'weekly' | 'monthly' | 'yearly',
  interval: number,
): number {
  const ONE_DAY_MS = 24 * 60 * 60 * 1000;
  const ONE_HOUR_MS = 60 * 60 * 1000;
  const ONE_MINUTE_MS = 60 * 1000;

  switch (frequency) {
    case 'minutes':
      return interval * ONE_MINUTE_MS;
    case 'hours':
      return interval * ONE_HOUR_MS;
    case 'daily':
      return interval * ONE_DAY_MS;
    case 'weekly':
      return interval * 7 * ONE_DAY_MS;
    case 'monthly':
      // Approximation: 30 days
      return interval * 30 * ONE_DAY_MS;
    case 'yearly':
      // Approximation: 365 days
      return interval * 365 * ONE_DAY_MS;
  }
}

File: crates/tasklens-core/src/utils/time.rs
use crate::types::Frequency;
use std::time::{SystemTime, UNIX_EPOCH};

/// Returns the current timestamp in milliseconds since the Unix Epoch.
pub fn get_current_timestamp() -> f64 {
    #[cfg(target_arch = "wasm32")]
    {
        js_sys::Date::now()
    }
    #[cfg(not(target_arch = "wasm32"))]
    {
        SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .expect("Time went backwards")
            .as_millis() as f64
    }
}

pub const DEFAULT_TASK_LEAD_TIME_MS: f64 = 7.0 * 24.0 * 60.0 * 60.0 * 1000.0;

/// Helper: Convert frequency/interval to milliseconds.
/// Note: simplified logic for MVP (ignoring leap years/DST nuances for basic "days/weeks").
pub fn get_interval_ms(frequency: Frequency, interval: f64) -> f64 {
    const ONE_MINUTE_MS: f64 = 60.0 * 1000.0;
    const ONE_HOUR_MS: f64 = 60.0 * ONE_MINUTE_MS;
    const ONE_DAY_MS: f64 = 24.0 * ONE_HOUR_MS;

    match frequency {
        Frequency::Minutes => interval * ONE_MINUTE_MS,
        Frequency::Hours => interval * ONE_HOUR_MS,
        Frequency::Daily => interval * ONE_DAY_MS,
        Frequency::Weekly => interval * 7.0 * ONE_DAY_MS,
        Frequency::Monthly => interval * 30.0 * ONE_DAY_MS,
        Frequency::Yearly => interval * 365.0 * ONE_DAY_MS,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_get_interval_ms() {
        assert_eq!(get_interval_ms(Frequency::Minutes, 1.0), 60_000.0);
        assert_eq!(get_interval_ms(Frequency::Hours, 1.0), 3_600_000.0);
        assert_eq!(get_interval_ms(Frequency::Daily, 1.0), 86_400_000.0);
        assert_eq!(get_interval_ms(Frequency::Weekly, 1.0), 604_800_000.0);
        assert_eq!(get_interval_ms(Frequency::Monthly, 1.0), 2_592_000_000.0);
        assert_eq!(get_interval_ms(Frequency::Yearly, 1.0), 31_536_000_000.0);
    }
}
